internal/server/data/migrations.go

```diff
@@ -655,3 +657,135 @@ func cleanCrossOrgGroupMemberships() *migrator.Migration {
 		},
 	}
 }
+
+func addProviderGroups() *migrator.Migration {
+	return &migrator.Migration{
+		ID: "2022-09-22T13:00",
+		Migrate: func(tx migrator.DB) error {
+			// setup the new tables
+			_, err := tx.Exec(
+				`
+					CREATE TABLE IF NOT EXISTS provider_groups (
+						organization_id bigint NOT NULL,
+						provider_id bigint NOT NULL,
+						name text NOT NULL,
+						created_at timestamp with time zone,
+						updated_at timestamp with time zone
+					);
+					CREATE TABLE IF NOT EXISTS provider_groups_provider_users (
+						provider_id bigint NOT NULL,
+						provider_group_name text NOT NULL,
+						provider_user_identity_id bigint NOT NULL
+					);
+					CREATE TABLE IF NOT EXISTS provider_groups_groups (
+						provider_id bigint NOT NULL,
+						provider_group_name text NOT NULL,
+						group_id bigint NOT NULL
+					);
+					CREATE UNIQUE INDEX idx_provider_group_names ON provider_groups(organization_id, provider_id, name);
+					CREATE UNIQUE INDEX idx_provider_groups_provider_users ON provider_groups_provider_users(provider_id, provider_group_name, provider_user_identity_id);
+					CREATE UNIQUE INDEX idx_provider_groups_groups ON provider_groups_groups(provider_id, provider_group_name, group_id);
+					
+					-- create a resolved_identity_groups view that can be used to check if a user is a direct member of a group,
+					-- or a user is a member of an external identity provider group that is mapped to this group in Infra
+					CREATE VIEW resolved_identity_groups AS
+						SELECT DISTINCT * FROM identities_groups
```

Comment:
I did not realize we would still have an `identities_group` table. I'm assuming this is for the infra provider, is that right?

Do you know if it would be possible to manage the infra provider like we do other providers? Would that be a large change?

I suspect the full join here could be expensive, so if we have an approach to remove it (either in this PR or a follow up) that might be worth exploring.

Reply:
I took another look at this and ended up adding all group relations to the `identites_groups` table. The Infra groups get added as "from the Infra provider".